pkg/distributor/ratestore.go

```diff
@@ -1,25 +1,200 @@
+					Buckets:   prometheus.DefBuckets,
+				}, instrument.HistogramCollectorBuckets,
+			),
+		),
+	}
+
+	s.Service = services.
+		NewTimerService(s.rateCollectionInterval, s.instrumentedUpdateAllRates, s.instrumentedUpdateAllRates, nil).
+		WithName("rate store")
+
+	return s
+}
+
+func (s *rateStore) instrumentedUpdateAllRates(ctx context.Context) error {
+	return instrument.CollectedRequest(ctx, "GetAllStreamRates", s.refreshDuration, instrument.ErrorCode, s.updateAllRates)
+}
+
+func (s *rateStore) updateAllRates(ctx context.Context) error {
+	clients, err := s.getClients()
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting ingester clients", "err", err)
+		s.rateRefreshFailures.WithLabelValues("ring_error").Inc()
+		return nil // Don't fail the service because we have an error getting the clients once
+	}
+
+	streamRates := s.getRates(ctx, clients)
+	rates := s.aggregateByShard(streamRates)
+
+	s.rateLock.Lock()
+	defer s.rateLock.Unlock()
+	s.rates = rates
+
+	return nil
 }
 
-func (u *unshardableStreamErr) Error() string {
-	return fmt.Sprintf("couldn't shard stream %s. number of shards (%d) is higher than number of entries (%d)", u.labels, u.shardNum, u.entriesNum)
+func (s *rateStore) aggregateByShard(streamRates map[uint64]*logproto.StreamRate) map[uint64]int64 {
+	rates := make(map[uint64]int64)
+	for _, sr := range streamRates {
+		if _, ok := rates[sr.StreamHashNoShard]; ok {
+			rates[sr.StreamHashNoShard] += sr.Rate
+			continue
+		}
+
+		rates[sr.StreamHashNoShard] = sr.Rate
+	}
+	return rates
 }
 
-type noopRateStore struct {
-	rate int
+func (s *rateStore) getRates(ctx context.Context, clients []logproto.StreamDataClient) map[uint64]*logproto.StreamRate {
+	parallelClients := make(chan logproto.StreamDataClient, len(clients))
+	responses := make(chan *logproto.StreamRatesResponse, len(clients))
+
+	for i := 0; i < s.maxParallelism; i++ {
+		go s.getRatesFromIngesters(ctx, parallelClients, responses)
+	}
+
+	for _, c := range clients {
+		parallelClients <- c
+	}
+	close(parallelClients)
+
+	return ratesPerStream(responses, len(clients))
+}
+
+func ratesPerStream(responses chan *logproto.StreamRatesResponse, totalResponses int) map[uint64]*logproto.StreamRate {
+	streamRates := make(map[uint64]*logproto.StreamRate)
+	for i := 0; i < totalResponses; i++ {
+		resp := <-responses
+		if resp == nil {
+			continue
+		}
+
+		for i := 0; i < len(resp.StreamRates); i++ {
+			rate := resp.StreamRates[i]
+
+			if r, ok := streamRates[rate.StreamHash]; ok {
+				if r.Rate < rate.Rate {
+					streamRates[rate.StreamHash] = rate
+				}
+				continue
+			}
+
+			streamRates[rate.StreamHash] = rate
+		}
+	}
+
+	return streamRates
 }
 
-func (n *noopRateStore) RateFor(stream *logproto.Stream) (int, error) {
-	return n.rate, nil
+func (s *rateStore) getRatesFromIngesters(ctx context.Context, clients chan logproto.StreamDataClient, responses chan *logproto.StreamRatesResponse) {
```

Comment:
thats extremely nitpicking but do you mind having this function before `ratesPerStream`? to align with the order of invocation